lib/std/math.zig

```diff
@@ -1484,3 +1484,66 @@ test "boolMask" {
 pub fn comptimeMod(num: anytype, denom: comptime_int) IntFittingRange(0, denom - 1) {
     return @intCast(IntFittingRange(0, denom - 1), @mod(num, denom));
 }
+
+/// returns -1, 0 or 1.  supports integer types, vectors of integer types, and float types
+/// - unsigned integer types will always return a 0 or 1
+/// TODO: support vectors of floats
+pub fn sign(i: anytype) @TypeOf(i) {
+    const T = @TypeOf(i);
+    return switch (@typeInfo(T)) {
+        .Int, .ComptimeInt => @as(T, @boolToInt(i > 0)) - @boolToInt(i < 0),
+        .Float, .ComptimeFloat => @intToFloat(T, @boolToInt(i > 0)) - @intToFloat(T, @boolToInt(i < 0)),
+        .Vector => |vinfo| blk: {
+            const u1xN = std.meta.Vector(vinfo.len, u1);
+            break :blk switch (@typeInfo(vinfo.child)) {
+                .Int => @as(T, @bitCast(u1xN, i > @splat(vinfo.len, @as(vinfo.child, 0)))) -
+                    @as(T, @bitCast(u1xN, i < @splat(vinfo.len, @as(vinfo.child, 0)))),
+                .Float => @compileError("TODO: add support for vectors of floats once @intToFloat accepts vector types"),
+                // break :blk @intToFloat(T, @bitCast(u1xN, i > @splat(vinfo.len, @as(vinfo.child, 0)))) -
+                //     @intToFloat(T, @bitCast(u1xN, i < @splat(vinfo.len, @as(vinfo.child, 0)))),
+                else => @compileError("Expected vector of ints or floats, found " ++ @typeName(T)),
+            };
+        },
+        else => @compileError("Expected an int, float or vector of one, found " ++ @typeName(T)),
+    };
+}
+
+test "sign" {
+    inline for ([_]type{ i8, i32, i64, u8, f16, f32, f64 }) |T| {
+        const input_expecteds = [_][2]comptime_int{
+            .{ 2, 1 },
+            .{ 0, 0 },
+        };
+        inline for (input_expecteds) |in_ex| {
+            const input = in_ex[0];
+            const expected = in_ex[1];
+            {
+                var actual = sign(@as(T, input));
+                try std.testing.expectEqual(@as(T, expected), actual);
+            }
+            const tinfo = @typeInfo(T);
+            // don't check negatives when T is an unsigned int
+            const neg = if (tinfo == .Int and tinfo.Int.signedness == .unsigned) 1 else -1;
+            {
+                var actual = sign(@as(T, neg * input));
+                try std.testing.expectEqual(@as(T, neg * expected), actual);
+            }
+            if (tinfo != .Float) { // TODO remove this check once @intToFloat accepts vectors
+                const uTx2 = std.meta.Vector(2, T);
+                var actual = sign(uTx2{ input, neg * input });
+                try std.testing.expectEqual(
+                    uTx2{ expected, neg * expected },
+                    actual,
+                );
+            }
+        }
+    }
+    comptime {
+        try std.testing.expectEqual(-1, sign(-10));
+        try std.testing.expectEqual(1, sign(10));
+        try std.testing.expectEqual(0, sign(0));
```

Comment:
If you want to add more edge cases:
Add -0.0, 0.0 for floats (yes -0.0 is a valid floating point number), -inf, inf.
If you want to go the extra mile, add 2 denormal numbers (+ and -), but it should be fine without.